R/PipeOpTrafo.R

```diff
@@ -0,0 +1,240 @@
+      # return a modified task and set the state
+      # make use of private$update_target
+      # therefore, typically calculate new_target (name it appropriately), do task$cbind(new_target)
+      # and provide names(new_target) as a character to update_target, e.g.:
+      #
+      # self$state = list()
+      # new_target = foo
+      # names(new_target) = "foo"
+      # task$cbind(new_target)
+      # private$update_target(task, new_target = names(new_target), new_type = NULL, ...)
+      stop("Abstract.")
+    },
+
+    train_invert = function(task) {
+      # return a predict_phase_control object (can be anything)
+      stop("Abstract.")
+    },
+
+    inverter = function(prediction, predict_phase_control) {
+      # function that inverts the predictions and returns a Prediction object
+      stop("Abstract.")
+    }
+  ),
+  private = list(
+    update_target = function(task, new_target, new_type = NULL, ...) {
+      type = if (is.null(new_type)) task$task_type else assert_subset(new_type, mlr_reflections$task_types$type, empty.ok = FALSE)
+      get(mlr_reflections$task_types[type]$task)$new(id = task$id, backend = task$backend, target = new_target, ...)
+    },
+
+    invert_help = function(predict_phase_control) {
+      function(inputs) {
+        assert_list(inputs, len = 1L, types = "Prediction")
+        self$inverter(inputs[[1L]], predict_phase_control)
+      }
+    },
+
+    set_target_to_NA_and_possibly_convert_task_type = function(task) {
+      #FIXME: convert the task target to NA and convert the task type?
+      # or just call train_target again?
+      intask = task$clone(deep = TRUE)
+      self$train_target(intask)
+    }
+  )
+)
+
+#' @title PipeOpInverter
+#'
+#' @usage NULL
+#' @name mlr_pipeops_inverter
+#' @format [`R6Class`] object inheriting from [`PipeOp`].
+#'
+#' @description
+#' #FIXME:
+#'
+#' @section Construction:
+#' #FIXME:
+#' ```
+#' PipeOpInverter$new(id)
+#' ```
+#'
+#' * `id` :: `character(1)`\cr
+#'   Identifier of resulting object. See `$id` slot of [`PipeOp`].
+#'
+#' @section Input and Output Channels:
+#' #FIXME:
+#'
+#' @section State:
+#' #FIXME:
+#'
+#' @section Internals:
+#' #FIXME:
+#'
+#' @section Methods:
+#' Methods inherited from [`PipeOp`], as well as:
+#' #FIXME:
+#'
+#' @family PipeOps
+#' @include PipeOp.R
+#' @export
+PipeOpInverter = R6Class("PipeOpInverter",
+  inherit = PipeOp,
+  public = list(
+    initialize = function(id = "inverter") {
+      super$initialize(id = id,
+        input = data.table(name = c("fun", "prediction"), train = c("NULL", "NULL"), predict = c("function", "Prediction")),
+        output = data.table(name = "output", train = "NULL", predict = "Prediction")
+      )
+    },
+
+    train_internal = function(inputs) {
+      self$state = list()
+      list(NULL)
+    },
+
+    predict_internal = function(inputs) {
+      list(inputs[[1L]](inputs[-1L]))
+    }
+  )
+)
+
+mlr_pipeops$add("inverter", PipeOpInverter)
+
+#' @title PipeOpLogtrafo
```

Comment:
PipeOpSimpleTrafo:
- following params in its `param_set`: `trafo`, `inverter`, `new_target_name`, `new_task_type`
- `trafo` and `inverter` default to `identity`
- `new_target_name` defaults to `paste0("trafo_", xxold_target_name)`
- `new_task_type` defaults to `NULL` (keep old task type)

Reply:
then `PipeOpLogTrafo` could inherit from this and just define different `trafo` and `inverter`.

Reply:
done, maybe we discuss whether another inheritance from `PipeOpSimpleTrafo` is needed for `PipeOpLogTrafo` (it is pretty easy to set this up already in `PipeOpSimpleTrafo`)